beacon-chain/core/electra/deposits.go

```diff
@@ -2,15 +2,248 @@ package electra
+	pubKey := data.PublicKey
+	amount := data.Amount
+	withdrawalCredentials := data.WithdrawalCredentials
+	index, ok := beaconState.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubKey))
+	if !ok {
+		if verifySignature {
+			valid, err := IsValidDepositSignature(data)
+			if err != nil {
+				return nil, err
+			}
+			if !valid {
+				return beaconState, nil
+			}
+		}
+		if err := AddValidatorToRegistry(beaconState, pubKey, withdrawalCredentials, amount); err != nil {
+			return nil, err
+		}
+	} else {
+		// phase0 bug: no validation on top ups. not something we need to fix here
+		if err := beaconState.AppendPendingBalanceDeposit(index, amount); err != nil {
+			return nil, err
+		}
+		if verifySignature {
+			valid, err := IsValidDepositSignature(data)
+			if err != nil {
+				return nil, err
+			}
+			if !valid {
+				return beaconState, nil
+			}
+		}
+		val, err := beaconState.ValidatorAtIndex(index)
+		if err != nil {
+			return nil, err
+		}
+		if helpers.IsCompoundingWithdrawalCredential(withdrawalCredentials) && helpers.HasETH1WithdrawalCredential(val) {
+			if err := SwitchToCompoundingValidator(beaconState, index); err != nil {
+				return nil, err
+			}
+		}
+	}
+	return beaconState, nil
+}
+
+// AddValidatorToRegistry updates the beacon state with validator information
+// def add_validator_to_registry(state: BeaconState,
+//
+//	                          pubkey: BLSPubkey,
+//	                          withdrawal_credentials: Bytes32,
+//	                          amount: uint64) -> None:
+//	index = get_index_for_new_validator(state)
+//	validator = get_validator_from_deposit(pubkey, withdrawal_credentials)
+//	set_or_append_list(state.validators, index, validator)
+//	set_or_append_list(state.balances, index, 0)  # [Modified in Electra:EIP7251]
+//	set_or_append_list(state.previous_epoch_participation, index, ParticipationFlags(0b0000_0000))
+//	set_or_append_list(state.current_epoch_participation, index, ParticipationFlags(0b0000_0000))
+//	set_or_append_list(state.inactivity_scores, index, uint64(0))
+//	state.pending_balance_deposits.append(PendingBalanceDeposit(index=index, amount=amount))  # [New in Electra:EIP7251]
+func AddValidatorToRegistry(beaconState state.BeaconState, pubKey []byte, withdrawalCredentials []byte, amount uint64) error {
+	val := GetValidatorFromDeposit(pubKey, withdrawalCredentials)
+	if err := beaconState.AppendValidator(val); err != nil {
+		return err
+	}
+	index, ok := beaconState.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubKey))
+	if !ok {
+		return errors.New("could not find validator in registry")
+	}
+	if err := beaconState.AppendBalance(0); err != nil {
+		return err
+	}
+	if err := beaconState.AppendPendingBalanceDeposit(index, amount); err != nil {
+		return err
+	}
+	if err := beaconState.AppendInactivityScore(0); err != nil {
+		return err
+	}
+	if err := beaconState.AppendPreviousParticipationBits(0); err != nil {
+		return err
+	}
+	return beaconState.AppendCurrentParticipationBits(0)
+}
```

Comment:
Shouldn't this go in the validator file?

Reply:
I can move it, i think it's only used in deposits though